File: packages/gavio-js/src/interceptors/pii/guard.ts
/**
 * PiiGuard — the pre/post interceptor that detects and redacts PII.
 *
 * Pipeline rule (privacy): PII is scanned on every request before it reaches
 * the provider. Detected entities are redacted/masked/tagged or blocked. In
 * REDACT mode the original values are restored in the response.
 */

import type { InterceptorContext } from '../../context.js'
import { PiiBlockedError } from '../../errors.js'
import type { GavioRequest } from '../../request.js'
import type { GavioResponse } from '../../response.js'
import { PiiMode, Sensitivity } from '../../types.js'
import type { Message } from '../../types.js'
import type { Interceptor } from '../base.js'
import { ScanContext } from './context.js'
import { matchLength } from './match.js'
import type { PiiMatch } from './match.js'
import { scannerTier } from './scanner.js'
import type { PiiScanner } from './scanner.js'
import { defaultScanners } from './scanners/index.js'

const STATE_KEY = 'pii_replacements'

// Confidence floor per sensitivity level — matches below the floor are ignored.
const CONFIDENCE_FLOOR: Record<Sensitivity, number> = {
  [Sensitivity.STRICT]: 0.0,
  [Sensitivity.BALANCED]: 0.6,
  [Sensitivity.PERMISSIVE]: 0.9,
}

export interface PiiGuardOptions {
  scanners?: PiiScanner[]
  sensitivity?: Sensitivity
  mode?: PiiMode
  restoreOnResponse?: boolean
  logEntityTypes?: boolean
  dryRun?: boolean
  locale?: string
  language?: string
}

class PiiGuard implements Interceptor {
  readonly name = 'pii_guard'
  readonly dryRunSafe = true

  private readonly scanners: PiiScanner[]
  private readonly sensitivity: Sensitivity
  private readonly mode: PiiMode
  private readonly restoreOnResponse: boolean
  private readonly logEntityTypes: boolean
  private readonly ownDryRun: boolean
  private readonly locale: string
  private readonly language: string

  constructor(options: PiiGuardOptions = {}) {
    this.scanners = options.scanners ?? defaultScanners()
    this.sensitivity = options.sensitivity ?? Sensitivity.STRICT
    this.mode = options.mode ?? PiiMode.REDACT
    this.restoreOnResponse = options.restoreOnResponse ?? true
    this.logEntityTypes = options.logEntityTypes ?? true
    this.ownDryRun = options.dryRun ?? false
    this.locale = options.locale ?? 'NL'
    this.language = options.language ?? 'en'
  }

  async before(request: GavioRequest, ctx: InterceptorContext): Promise<GavioRequest> {
    const scanCtx = new ScanContext(this.language, this.locale)
    const floor = CONFIDENCE_FLOOR[this.sensitivity]

    const newMessages: Message[] = []
    const allTypes: string[] = []
    const replacements: Record<string, string> =
      (ctx.state[STATE_KEY] as Record<string, string> | undefined) ?? {}

    const isDryRun = this.ownDryRun || ctx.dryRun

    for (const message of request.messages) {
      const content = message.content ?? ''
      const matches = await this.scanText(content, scanCtx, floor)
      for (const m of matches) allTypes.push(m.entityType)

      if (matches.length > 0 && this.mode === PiiMode.BLOCK) {
        const types = matches.map((m) => m.entityType)
        throw new PiiBlockedError(types)
      }

      let redacted = content
      if (matches.length > 0 && !isDryRun) {
        redacted = this.apply(content, matches, replacements)
      }

      newMessages.push({ ...message, content: redacted })
    }

    if (allTypes.length > 0) {
      ctx.recordPii(allTypes)
      if (this.logEntityTypes) {
        const unique = Array.from(new Set(allTypes)).sort()
        // eslint-disable-next-line no-console
        console.info(`[gavio:pii] detected entity types: ${unique.join(', ')}`)
      }
    }

    if (this.restoreOnResponse && Object.keys(replacements).length > 0) {
      ctx.state[STATE_KEY] = replacements
    }

    if (isDryRun) return request
    return request.copyWithMessages(newMessages)
  }

  async after(
    response: GavioResponse,
    ctx: InterceptorContext,
  ): Promise<GavioResponse> {
    if (!this.restoreOnResponse || this.mode !== PiiMode.REDACT) return response
    const replacements = ctx.state[STATE_KEY] as Record<string, string> | undefined
    if (!replacements || Object.keys(replacements).length === 0) return response
    let content = response.content
    for (const [token, original] of Object.entries(replacements)) {
      content = content.split(token).join(original)
    }
    if (content === response.content) return response
    return response.copyWithContent(content)
  }

  private async scanText(
    text: string,
    scanCtx: ScanContext,
    floor: number,
  ): Promise<PiiMatch[]> {
    const raw: PiiMatch[] = []
    const ordered = [...this.scanners].sort((a, b) => scannerTier(a) - scannerTier(b))
    for (const scanner of ordered) {
      const found = await scanner.scan(text, scanCtx)
      for (const match of found) {
        if (match.confidence >= floor) raw.push(match)
      }
    }
    return resolveOverlaps(raw)
  }

  private apply(
    text: string,
    matches: PiiMatch[],
    replacements: Record<string, string>,
  ): string {
    // Replace right-to-left so earlier offsets stay valid.
    const ordered = [...matches].sort((a, b) => b.start - a.start)
    let out = text
    for (const match of ordered) {
      const token = this.tokenFor(match)
      if (this.mode === PiiMode.REDACT) {
        replacements[token] = match.value
      }
      out = out.slice(0, match.start) + token + out.slice(match.end)
    }
    return out
  }

  private tokenFor(match: PiiMatch): string {
    if (this.mode === PiiMode.MASK) {
      return '*'.repeat(Math.max(matchLength(match), 1))
    }
    if (this.mode === PiiMode.TAG) {
      return `<${match.entityType}>${match.value}</${match.entityType}>`
    }
    // REDACT (default)
    return match.replacement || `[${match.entityType}]`
  }
}

/**
 * Drop lower-priority matches that overlap a kept one.
 *
 * Sort by start, then by descending span length (prefer the longer match),
 * then by confidence. Greedily keep non-overlapping matches.
 */
export function resolveOverlaps(matches: PiiMatch[]): PiiMatch[] {
  const ordered = [...matches].sort((a, b) => {
    if (a.start !== b.start) return a.start - b.start
    const lenDiff = matchLength(b) - matchLength(a)
    if (lenDiff !== 0) return lenDiff
    return b.confidence - a.confidence
  })
  const kept: PiiMatch[] = []
  let occupiedEnd = -1
  for (const match of ordered) {
    if (match.start >= occupiedEnd) {
      kept.push(match)
      occupiedEnd = match.end
    }
  }
  return kept
}

/** Factory: build a PiiGuard interceptor. */
export function piiGuard(options: PiiGuardOptions = {}): Interceptor {
  return new PiiGuard(options)
}

File: packages/gavio-py/gavio/interceptors/pii/guard.py
"""PiiGuard — the pre/post interceptor that detects and redacts PII.

Pipeline rule (MASTER_PLAN P5 / privacy): PII is scanned on every request
before it reaches the provider. Detected entities are redacted/masked/tagged
or blocked. In REDACT mode the original values are restored in the response.
"""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING, Any

from ...context import InterceptorContext
from ...exceptions import PiiBlockedError
from ...request import GavioRequest
from ...response import GavioResponse
from ...types import PiiMode, Sensitivity
from ..base import Interceptor
from .context import ScanContext
from .match import PiiMatch
from .scanner import PiiScanner
from .scanners import default_scanners

if TYPE_CHECKING:
    from .policy_pack import PolicyPack

logger = logging.getLogger("gavio.pii")

_STATE_KEY = "pii_replacements"

# Confidence floor per sensitivity level — matches below the floor are ignored.
_CONFIDENCE_FLOOR = {
    Sensitivity.STRICT: 0.0,
    Sensitivity.BALANCED: 0.6,
    Sensitivity.PERMISSIVE: 0.9,
}


class PiiGuard(Interceptor):
    """Scan request messages for PII and act per the configured mode."""

    def __init__(
        self,
        scanners: list[PiiScanner] | None = None,
        sensitivity: Sensitivity | str = Sensitivity.STRICT,
        mode: PiiMode | str = PiiMode.REDACT,
        restore_on_response: bool = True,
        log_entity_types: bool = True,
        dry_run: bool = False,
        locale: str = "NL",
        language: str = "en",
    ) -> None:
        self._scanners = scanners if scanners is not None else default_scanners()
        self.sensitivity = Sensitivity(sensitivity)
        self.mode = PiiMode(mode)
        self.restore_on_response = restore_on_response
        self.log_entity_types = log_entity_types
        self._dry_run = dry_run
        self.locale = locale
        self.language = language

    @property
    def name(self) -> str:
        return "pii_guard"

    @property
    def dry_run_safe(self) -> bool:
        return True

    @classmethod
    def from_policy_pack(cls, *packs: PolicyPack, **kwargs: Any) -> PiiGuard:
        from .policy_pack import policy_pack_scanners

        return cls(scanners=policy_pack_scanners(*packs), **kwargs)

    async def before(
        self, request: GavioRequest, ctx: InterceptorContext
    ) -> GavioRequest:
        scan_ctx = ScanContext(language=self.language, locale=self.locale)
        floor = _CONFIDENCE_FLOOR[self.sensitivity]

        new_messages: list[dict[str, str]] = []
        all_types: list[str] = []
        replacements: dict[str, str] = ctx.state.get(_STATE_KEY, {})

        for message in request.messages:
            content = message.get("content", "")
            matches = self._scan_text(content, scan_ctx, floor)
            all_types.extend(m.entity_type for m in matches)

            if matches and self.mode == PiiMode.BLOCK:
                types = [m.entity_type for m in matches]
                logger.warning("pii_guard BLOCK: %s", sorted(set(types)))
                raise PiiBlockedError(types)

            redacted = content
            if matches and not (self._dry_run or ctx.dry_run):
                redacted = self._apply(content, matches, replacements)

            new_msg = dict(message)
            new_msg["content"] = redacted
            new_messages.append(new_msg)

        if all_types:
            ctx.record_pii(all_types)
            if self.log_entity_types:
                logger.info(
                    "pii_guard detected entity types: %s", sorted(set(all_types))
                )

        if self.restore_on_response and replacements:
            ctx.state[_STATE_KEY] = replacements

        if self._dry_run or ctx.dry_run:
            return request
        return request.copy_with_messages(new_messages)

    async def after(
        self, response: GavioResponse, ctx: InterceptorContext
    ) -> GavioResponse:
        if not self.restore_on_response or self.mode != PiiMode.REDACT:
            return response
        replacements: dict[str, str] = ctx.state.get(_STATE_KEY, {})
        if not replacements:
            return response
        content = response.content
        # Replace placeholder tokens back with original values.
        for token, original in replacements.items():
            content = content.replace(token, original)
        if content == response.content:
            return response
        return response.copy_with_content(content)

    def _scan_text(
        self, text: str, scan_ctx: ScanContext, floor: float
    ) -> list[PiiMatch]:
        raw: list[PiiMatch] = []
        for scanner in sorted(self._scanners, key=lambda s: s.tier):
            for match in scanner.scan(text, scan_ctx):
                if match.confidence >= floor:
                    raw.append(match)
        return _resolve_overlaps(raw)

    def _apply(
        self,
        text: str,
        matches: list[PiiMatch],
        replacements: dict[str, str],
    ) -> str:
        # Replace right-to-left so earlier offsets stay valid.
        for match in sorted(matches, key=lambda m: m.start, reverse=True):
            token = self._token_for(match)
            if self.mode == PiiMode.REDACT:
                replacements[token] = match.value
            text = text[: match.start] + token + text[match.end :]
        return text

    def _token_for(self, match: PiiMatch) -> str:
        if self.mode == PiiMode.MASK:
            return "*" * max(match.length, 1)
        if self.mode == PiiMode.TAG:
            return f"<{match.entity_type}>{match.value}</{match.entity_type}>"
        # REDACT (default)
        return match.replacement or f"[{match.entity_type}]"


def _resolve_overlaps(matches: list[PiiMatch]) -> list[PiiMatch]:
    """Drop lower-priority matches that overlap a kept one.

    Sort by start, then by descending span length (prefer the longer match),
    then by confidence. Greedily keep non-overlapping matches.
    """
    ordered = sorted(matches, key=lambda m: (m.start, -m.length, -m.confidence))
    kept: list[PiiMatch] = []
    occupied_end = -1
    for match in ordered:
        if match.start >= occupied_end:
            kept.append(match)
            occupied_end = match.end
    return kept
